homeassistant/components/lutron_caseta/__init__.py

```diff
@@ -218,57 +221,153 @@ def _async_register_bridge_device(
 
 
 @callback
-def _async_register_button_devices(
+def _async_setup_keypads(
     hass: HomeAssistant,
     config_entry_id: str,
     bridge: Smartbridge,
     bridge_device: dict[str, Any],
-    button_devices_by_id: dict[int, dict],
-) -> tuple[dict[str, dict], dict[int, DeviceInfo]]:
-    """Register button devices (Pico Remotes) in the device registry."""
+) -> LutronKeypadData:
+    """Register keypad devices (Keypads and Pico Remotes) in the device registry."""
+
     device_registry = dr.async_get(hass)
-    button_devices_by_dr_id: dict[str, dict] = {}
-    device_info_by_device_id: dict[int, DeviceInfo] = {}
-    seen: set[str] = set()
+
     bridge_devices = bridge.get_devices()
+    bridge_buttons = bridge.buttons
+
+    dr_device_id_to_keypad: dict[str, LutronKeypad] = {}
+    keypads: dict[int, LutronKeypad] = {}
+    keypad_buttons: dict[int, LutronButton] = {}
+    keypad_button_names_to_leap: dict[int, dict[str, int]] = {}
+
+    keypad: LutronKeypad
+    for bridge_button in bridge_buttons.values():
+
+        bridge_keypad = bridge_devices[bridge_button["parent_device"]]
+        keypad_device_id = bridge_keypad["device_id"]
+        button_device_id = bridge_button["device_id"]
+
+        if keypad_device_id not in keypads:
+            # First time seeing this keypad, build keypad data and store in keypads
+            area_name = _area_name_from_id(bridge.areas, bridge_keypad["area"])
+            keypad_name = bridge_keypad["name"].split("_")[-1]
+            keypad_serial = _handle_none_keypad_serial(
+                bridge_keypad, bridge_device["serial"]
+            )
+            device_info = DeviceInfo(
+                name=f"{area_name} {keypad_name}",
+                manufacturer=MANUFACTURER,
+                identifiers={(DOMAIN, keypad_serial)},
+                model="{bridge_keypad['model']} ({bridge_keypad['type']})",
+                via_device=(DOMAIN, bridge_device["serial"]),
+            )
+            if area_name != UNASSIGNED_AREA:
+                device_info["suggested_area"] = area_name
 
-    for device in button_devices_by_id.values():
+            dr_device = device_registry.async_get_or_create(
+                **device_info, config_entry_id=config_entry_id
+            )
 
-        ha_device = device
-        if "parent_device" in device and device["parent_device"] is not None:
-            # Device is a child of parent_device
-            # use the parent_device for HA device info
-            ha_device = bridge_devices[device["parent_device"]]
+            keypad = keypads.setdefault(
```

Comment:
I'm a bit lost about why setdefault is being used here

Reply:
What method would use use to add the entry to the dict?

Reply:
I'd write it like this : https://github.com/bdraco/home-assistant/commit/f6bdcd9190ad9774abd2b89026f6462505309632 to avoid carrying the variable between loop scopes

Reply:
sure, I can do that

Reply:
addressed with commit https://github.com/home-assistant/core/pull/80797/commits/f787a1f4e10048ac2c59c63a35a581f5cfee0934